app/models/discussion_reader.rb

```diff
@@ -11,9 +11,10 @@ class DiscussionReader < ActiveRecord::Base
 
   def self.for(user: nil, discussion: nil)
     if user.is_logged_in?
-      where(user_id: user.id, discussion_id: discussion.id).first_or_initialize do |dr|
-        dr.discussion = discussion
-        dr.user = user
+      begin
```

Comment:
@gdpelican fyi: found this approach to making this thread safe:

http://www.rubydoc.info/docs/rails/ActiveRecord/Relation:find_or_create_by

See the bottom of the doc, above the comments